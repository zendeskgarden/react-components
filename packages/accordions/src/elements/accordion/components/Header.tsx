/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { useState, FocusEvent, forwardRef, HTMLAttributes } from 'react';
import { composeEventHandlers } from '@zendeskgarden/container-utilities';
import ChevronDown from '@zendeskgarden/svg-icons/src/16/chevron-down-stroke.svg';
import { useAccordionContext, useSectionContext, HeaderContext } from '../../../utils';
import { StyledHeader, StyledRotateIcon, COMPONENT_ID as buttonGardenId } from '../../../styled';

export const Header = forwardRef<HTMLDivElement, HTMLAttributes<HTMLDivElement>>((props, ref) => {
  const {
    level: ariaLevel,
    isCompact,
    isCollapsible,
    getHeaderProps,
    getTriggerProps,
    expandedSections
  } = useAccordionContext();
  const { onClick, onFocus, onBlur, onMouseOver, onMouseOut, children, ...other } = props;
  const sectionIndex = useSectionContext();
  const [isFocused, setIsFocused] = useState(false);
  const [isHovered, setIsHovered] = useState(false);
  const isExpanded = expandedSections.includes(sectionIndex);
  /**
   *  Pressing the space key on a button triggers the `onTriggerClick` callback.
   * `onKeyDown` is plucked out and not passed to the Label (button) element
   * to prevent double invocations of the click event.
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { onClick: onTriggerClick, onKeyDown, ...otherTriggerProps } = getTriggerProps({
    type: 'button',
    index: sectionIndex
  });
  const onHeaderFocus = (e: FocusEvent) => {
    e.persist();

    setTimeout(() => {
      const isAccordionButton = e.target.getAttribute('data-garden-id') === buttonGardenId;
      const isFocusVisible = e.target.getAttribute('data-garden-focus-visible');

      if (isAccordionButton && isFocusVisible) {
        setIsFocused(true);
      }
    }, 0);
  };

  const value = {
    isHovered,
    otherTriggerProps
  };

  return (
    <HeaderContext.Provider value={value}>
      <StyledHeader
        {...getHeaderProps({
          ref,
          ariaLevel,
          isCompact,
          isFocused,
          isExpanded,
          isCollapsible,
          onClick: composeEventHandlers(onClick, onTriggerClick),
          onFocus: composeEventHandlers(onFocus, onHeaderFocus),
          onBlur: composeEventHandlers(onBlur, () => setIsFocused(false)),
          onMouseOver: composeEventHandlers(onMouseOver, () => setIsHovered(true)),
          onMouseOut: composeEventHandlers(onMouseOut, () => setIsHovered(false)),
          ...other
        })}
      >
        {children}
        <StyledRotateIcon
          isCompact={isCompact}
          isHovered={isHovered}
          isRotated={isExpanded}
          isCollapsible={isCollapsible}
          onMouseOver={composeEventHandlers(onMouseOver, () => setIsHovered(true))}
          onMouseOut={composeEventHandlers(onMouseOut, () => setIsHovered(false))}
        >
          <ChevronDown />
        </StyledRotateIcon>
      </StyledHeader>
    </HeaderContext.Provider>
  );
});

Header.displayName = 'Header';
