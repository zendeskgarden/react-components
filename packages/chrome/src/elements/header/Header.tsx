/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { HTMLAttributes } from 'react';
import PropTypes from 'prop-types';
import { StyledHeader, IStyledHeaderProps } from '../../styled';

/**
 * @extends HTMLAttributes<HTMLElement>
 */
export const Header = React.forwardRef<
  HTMLElement,
  IStyledHeaderProps & HTMLAttributes<HTMLElement>
>((props, ref) => <StyledHeader ref={ref} {...props} />);

Header.displayName = 'Header';

Header.propTypes = {
  isStandalone: PropTypes.bool
};
