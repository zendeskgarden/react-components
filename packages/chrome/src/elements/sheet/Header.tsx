/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import { forwardRef, HTMLAttributes } from 'react';
import { StyledSheetHeader } from '../../styled';

export const SheetHeader = forwardRef<HTMLElement, HTMLAttributes<HTMLElement>>(({ children, ...props }, ref) => {
  return (
    <StyledSheetHeader ref={ref} {...props}>
      {children}
    </StyledSheetHeader>
  );
});
