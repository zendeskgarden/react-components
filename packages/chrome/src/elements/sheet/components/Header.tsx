/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { forwardRef, HTMLAttributes } from 'react';

import { StyledSheetHeader } from '../../../styled';
import { useSheetContext } from '../../../utils/useSheetContext';

const SheetHeader = forwardRef<HTMLElement, HTMLAttributes<HTMLElement>>((props, ref) => {
  const { isCloseButtonPresent } = useSheetContext();

  return <StyledSheetHeader ref={ref} isCloseButtonPresent={isCloseButtonPresent} {...props} />;
});

SheetHeader.displayName = 'Sheet.Header';

/**
 * @extends HTMLAttributes<HTMLElement>
 */
export const Header = SheetHeader;
