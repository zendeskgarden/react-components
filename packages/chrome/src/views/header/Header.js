/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import PropTypes from 'prop-types';
import styled from 'styled-components';
import classNames from 'classnames';
import { retrieveTheme } from '@zendeskgarden/react-theming';
import ChromeStyles from '@zendeskgarden/css-chrome';

const COMPONENT_ID = 'chrome.header';

/**
 * Accepts all `<header>` props
 */
const Header = styled.header.attrs({
  'data-garden-id': COMPONENT_ID,
  'data-garden-version': PACKAGE_VERSION,
  className: props =>
    classNames(ChromeStyles['c-chrome__body__header'], {
      [ChromeStyles['c-chrome__body__header--standalone']]: props.standalone
    })
})`
  ${props => retrieveTheme(COMPONENT_ID, props)};
`;

Header.propTypes = {
  /** Display logo for standlone usage  */
  standalone: PropTypes.bool
};

/** @component */
export default Header;
