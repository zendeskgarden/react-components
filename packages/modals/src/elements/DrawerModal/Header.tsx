/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import { useModalContext } from '../../utils/useModalContext';
import { StyledDrawerModalHeader } from '../../styled';

/**
 * @extends HTMLAttributes<HTMLDivElement>
 */
export const Header = React.forwardRef<HTMLDivElement, React.HTMLAttributes<HTMLDivElement>>(
  (props, ref) => {
    const { getTitleProps } = useModalContext();

    return <StyledDrawerModalHeader ref={ref} {...getTitleProps(props)} />;
  }
);

Header.displayName = 'DrawerModal.Header';
