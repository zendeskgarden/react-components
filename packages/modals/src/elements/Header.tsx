/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import { StyledHeader, IStyledHeaderProps } from '../styled';
import { useModalContext } from '../utils/useModalContext';

/**
 * Accepts all `<div>` props
 */
export const Header = React.forwardRef<
  HTMLDivElement,
  IStyledHeaderProps & React.HTMLAttributes<HTMLDivElement>
>((props, ref) => {
  const { getTitleProps } = useModalContext();

  return <StyledHeader ref={ref} {...getTitleProps(props)} />;
});
