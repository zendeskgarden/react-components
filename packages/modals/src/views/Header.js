/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import styled from 'styled-components';
import classNames from 'classnames';
import { retrieveTheme } from '@zendeskgarden/react-theming';
import ModalStyles from '@zendeskgarden/css-modals';

const COMPONENT_ID = 'modals.header';

/**
 * Accepts all `<div>` props
 */
const Header = styled.div.attrs(props => ({
  'data-garden-id': COMPONENT_ID,
  'data-garden-version': PACKAGE_VERSION,
  className: classNames(ModalStyles['c-dialog__header'], {
    // Danger styling
    [ModalStyles['c-dialog__header--danger']]: props.danger
  })
}))`
  ${props => retrieveTheme(COMPONENT_ID, props)};
`;

Header.hasType = () => Header;

/** @component */
export default Header;
