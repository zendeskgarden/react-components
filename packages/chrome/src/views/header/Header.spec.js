/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import { shallow } from 'enzyme';

import Header from './Header';

describe('Header', () => {
  it('renders default styling', () => {
    const wrapper = shallow(<Header />);

    expect(wrapper).toHaveClassName('c-chrome__body__header');
  });

  it('renders standalone styling if provided', () => {
    const wrapper = shallow(<Header standalone />);

    expect(wrapper).toHaveClassName('c-chrome__body__header--standalone');
  });
});
