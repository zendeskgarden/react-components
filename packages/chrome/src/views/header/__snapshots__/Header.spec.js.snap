// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders default styling 1`] = `
<header
  className="c-chrome__body__header "
  data-garden-id="chrome.header"
  data-garden-version="version"
/>
`;

exports[`Header renders standalone styling if provided 1`] = `
<header
  className="c-chrome__body__header c-chrome__body__header--standalone "
  data-garden-id="chrome.header"
  data-garden-version="version"
/>
`;
